rskj-core/src/main/java/org/ethereum/datasource/RocksDbDataSource.java

```diff
@@ -0,0 +1,374 @@
+            throw new RuntimeException(e);
+        }
+
+        resetDbLock.readLock().unlock();
+        profiler.stop(metric);
+    }
+
+    @Override
+    public Set<ByteArrayWrapper> keys() {
+        if (logger.isTraceEnabled()) {
+            logger.trace("~> RocksDbDataSource.keys(): {}", name);
+        }
+
+        Set<ByteArrayWrapper> result = new HashSet<>();
+
+        Metric metric = profiler.start(Profiler.PROFILING_TYPE.DB_READ);
+        resetDbLock.readLock().lock();
+
+        try (RocksIterator iterator = db.newIterator()) {
+
+            for (iterator.seekToFirst(); iterator.isValid(); iterator.next()) {
+                result.add(ByteUtil.wrap(iterator.key()));
+            }
+
+            if (logger.isTraceEnabled()) {
+                logger.trace("<~ RocksDbDataSource.keys(): {}, {}", name, result.size());
+            }
+        }
+
+        resetDbLock.readLock().unlock();
+        profiler.stop(metric);
+
+        return result;
+    }
+
+    private void updateBatchInternal(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> deleteKeys) {
+        Metric metric = profiler.start(Profiler.PROFILING_TYPE.DB_WRITE);
+        if (rows.containsKey(null) || rows.containsValue(null)) {
+            profiler.stop(metric);
+            throw new IllegalArgumentException("Cannot update null values");
+        }
+        // Note that this is not atomic.
+        try (WriteBatch batch = new WriteBatch()) {
+            for (Map.Entry<ByteArrayWrapper, byte[]> entry : rows.entrySet()) {
+                batch.put(entry.getKey().getData(), entry.getValue());
+            }
+            for (ByteArrayWrapper deleteKey : deleteKeys) {
+                batch.delete(deleteKey.getData());
+            }
+            db.write(new WriteOptions(), batch);
+            profiler.stop(metric);
+        } catch (RocksDBException e) {
+            logger.error("Exception. Not retrying.", e);
+        }
+
+    }
+
+    @Override
+    public void updateBatch(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> deleteKeys) {
+        if (rows.containsKey(null)) {
+            throw new IllegalArgumentException("Cannot update null values");
+        }
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("~> RocksDbDataSource.updateBatch(): {}, {}", name, rows.size());
+        }
+
+        resetDbLock.readLock().lock();
+
+        int retries = 0;
+        Exception exCaught = null;
+
+        while (retries < MAX_RETRIES) {
+            try {
+                updateBatchInternal(rows, deleteKeys);
+
+                if (logger.isTraceEnabled()) {
+                    logger.trace("<~ RocksDbDataSource.updateBatch(): {}, {}", name, rows.size());
+                }
+
+                break;
+            } catch (Exception e) {
+                logger.error("Exception. Retrying again...", e);
+                exCaught = e;
+            }
+
+            retries++;
+        }
+
+        resetDbLock.readLock().unlock();
```

Comment:
better to put this into the final block:
```
resetDbLock.readLock().unlock();
```

Reply:
Done